packages/tools/src/stateManagement/segmentation/helpers/clipAndCacheSurfacesForViewport.ts

```diff
@@ -0,0 +1,189 @@
+
+/**
+ * a cache from actorUID to cacheId to SurfaceClipResult
+ * Map<actorUID, Map<cacheId, SurfaceClipResult>>
+ * cacheId is slice specific (viewPlaneNormal, sliceIndex)
+ */
+const polyDataCache = new Map() as PolyDataClipCacheType;
+const surfacesAABBCache = new Map();
+
+const triggerWorkerProgress = (eventTarget, progress) => {
+  triggerEvent(eventTarget, Enums.Events.WEB_WORKER_PROGRESS, {
+    progress,
+    type: WorkerTypes.SURFACE_CLIPPING,
+  });
+};
+
+export async function clipAndCacheSurfacesForViewport(
+  surfacesInfo: SurfacesInfo[],
+  viewport: Types.IVolumeViewport,
+  segmentationRepresentationUID: string
+) {
+  registerPolySegWorker();
+  // All planes is an array of planes pairs for each slice, so we should loop over them and
+  // add the planes to the clipping filter and cache the results for that slice
+
+  // Fix these ts ignores
+  // @ts-ignore
+  const planesInfo = viewport.getSlicesClippingPlanes?.();
+
+  if (!planesInfo) {
+    // this means it is probably the stack viewport not being ready
+    // in terms of planes which we should wait for the first render to
+    // get the planes
+    return;
+  }
+
+  // @ts-ignore
+  const currentSliceIndex = viewport.getSliceIndex();
+
+  // Reorder planesInfo based on proximity to currentSliceIndex
+  planesInfo.sort((a, b) => {
+    const diffA = Math.abs(a.sliceIndex - currentSliceIndex);
+    const diffB = Math.abs(b.sliceIndex - currentSliceIndex);
+    return diffA - diffB;
+  });
+
+  triggerWorkerProgress(eventTarget, 0);
+
+  // check which surfaces don't have a cached AABB
+  // make a list of the surfaces that don't have a cached AABB
+  await updateSurfacesAABBCache(surfacesInfo);
+
+  const surfacesAABB = new Map();
+  surfacesInfo.forEach((surface) => {
+    surfacesAABB.set(surface.id, surfacesAABBCache.get(surface.id));
+  });
+
+  const camera = viewport.getCamera();
+
+  await workerManager
+    .executeTask(
+      'polySeg',
+      'cutSurfacesIntoPlanes',
+      {
+        surfacesInfo,
+        planesInfo,
+        surfacesAABB,
+      },
+      {
+        callbacks: [
+          // progress callback
+          ({ progress }) => {
+            triggerWorkerProgress(eventTarget, progress);
+          },
+          // update cache callback
+          ({ sliceIndex, polyDataResults }) => {
+            polyDataResults.forEach((polyDataResult, surfaceId) => {
+              const actorUID = `${segmentationRepresentationUID}_${surfaceId}`;
+              const cacheId = generateCacheId(
+                viewport,
+                camera.viewPlaneNormal,
+                sliceIndex
+              );
+              updatePolyDataCache(actorUID, cacheId, polyDataResult);
+            });
+          },
+        ],
+      }
+    )
+    .catch((error) => {
+      console.error(error);
+    });
+
+  triggerWorkerProgress(eventTarget, 1);
+
+  return polyDataCache;
+}
+
+async function updateSurfacesAABBCache(surfacesInfo: SurfacesInfo[]) {
```

Comment:
A simple comment would be nice - 
/** get the bounds for the surfaces contours into surfacesInfo as a cache of AABB values */